phpBB/phpbb/db/migration/data/v33x/bot_update.php

```diff
@@ -0,0 +1,91 @@
+<?php
+/**
+ *
+ * This file is part of the phpBB Forum Software package.
+ *
+ * @copyright (c) phpBB Limited <https://www.phpbb.com>
+ * @license GNU General Public License, version 2 (GPL-2.0)
+ *
+ * For full copyright and license information, please see
+ * the docs/CREDITS.txt file.
+ *
+ */
+
+namespace phpbb\db\migration\data\v33x;
+
+class bot_update extends \phpbb\db\migration\migration
+{
+	static public function depends_on()
+	{
+		return ['\phpbb\db\migration\data\v330\v330'];
+	}
+
+	public function update_data()
+	{
+		return [
+			['custom', [[&$this, 'add_duckduckgo_bot']]],
+		];
+	}
+
+	public function add_duckduckgo_bot()
+	{
+		$bot_name = 'DuckDuckGo [Bot]';
+		$bot_name_clean = utf8_clean_string($bot_name);
+
+		$sql = 'SELECT user_id
+			FROM ' . USERS_TABLE . '
+			WHERE ' . $this->db->sql_build_array('SELECT', ['username_clean' => $bot_name_clean]);
+		$result = $this->db->sql_query($sql);
+		$bot_exists = (bool) $this->db->sql_fetchfield('user_id');
+		$this->db->sql_freeresult($result);
+
+		if (!$bot_exists)
+		{
+			$bot_agent = 'DuckDuckBot/';
+			$bot_ip = '';
+			$sql = 'SELECT group_id, group_colour
+				FROM ' . GROUPS_TABLE . '
+				WHERE ' . $this->db->sql_build_array('SELECT', ['group_name' => 'BOTS']);
+			$result = $this->db->sql_query($sql);
+			$group_row = $this->db->sql_fetchrow($result);
+			$this->db->sql_freeresult($result);
+
+			// Default fallback, should never get here
+			if (!$group_row)
+			{
+				$group_row['group_id'] = 6;
+				$group_row['group_colour'] = '9E8DA7';
+			}
+
+			if (!function_exists('user_add'))
+			{
+				include($this->phpbb_root_path . 'includes/functions_user.' . $this->php_ext);
+			}
+
+			$user_row = [
+				'user_type'				=> USER_IGNORE,
+				'group_id'				=> $group_row['group_id'],
+				'username'				=> $bot_name,
+				'user_regdate'			=> time(),
+				'user_password'			=> '',
+				'user_colour'			=> $group_row['group_colour'],
+				'user_email'			=> '',
+				'user_lang'				=> $this->config['default_lang'],
+				'user_style'			=> $this->config['default_style'],
+				'user_timezone'			=> 0,
+				'user_dateformat'		=> $this->config['default_dateformat'],
+				'user_allow_massemail'	=> 0,
+			];
+
+			$user_id = user_add($user_row);
+			$sql = 'INSERT INTO ' . BOTS_TABLE . ' ' . $this->db->sql_build_array('INSERT', [
```

Comment:
What if the bots table does not exists? Edge case.
Not sure about the use of constants here, on a side note.

Reply:
It will fail, like any other migration where a column or table is missing. I would treat a missing bots table like a broken install.

I'm not sure about the constant either, but I was following the code style from recent migration files, like this one.

https://github.com/phpbb/phpbb/blob/master/phpBB/phpbb/db/migration/data/v330/remove_attachment_flash.php

I don't even like the final comma in the array elements, but that's in the coding guidelines.

Reply:
Preferably use the `$this->table_prefix` parameter which is already available. I know it isn't used in the other migration either, but we have to start somewhere. Also for the table constants above.

Reply:
> Preferably use the `$this->table_prefix` parameter which is already available. I know it isn't used in the other migration either, but we have to start somewhere. Also for the table constants above.

May I know where this discussion was made with regards to this?

Best regards 👍

Reply:
The idea is to move away from defined constants. So if there is an option to use the table prefix, ideally that should be used.

Reply:
So no actual discussion topic was made ?
Or it was made & decided that `$this->table_prefix` is to be used ?

Thanks for the responce.

Best regards 👍

Reply:
There are several discussions throughout, but there have been quite a few, not sure where they all are located. Table prefix is still not the ideal solution, that would be to use the tables parameter (the array). But that isn't added to 3.3 base migration yet, it is already in 4.0.

Reply:
Ahh! ok.

Thanks for your kind info. 😄 

Best regards 👍